### DISCUSSION TITLE
Avoiding rate limits when using pull-through cache for DockerHub

### DISCUSSION THREAD

**Discussion by erikschul** [distribution/distribution#4051] [Q&A]
The [documentation](https://docs.docker.com/registry/recipes/mirror/#what-if-the-content-changes-on-the-hub) says
```
When a pull is attempted with a tag, the Registry checks the remote to ensure if it has the latest version of the requested content. Otherwise, it fetches and caches the latest content.
```

but a major reason why I'm interested in using Distribution as a pull-through cache, is to avoid hitting the daily pull limit on DockerHub. As I understand it, retrieving the manifest (latest version) will still count as a pull?

Is there a way to avoid it? Perhaps using nginx with caching in front?

This seems especially relevant when using Kubernetes, as the CrashLoopBackOff (even when image has already been pulled) causes manifest downloads, which count against the daily limit, which can then cause a 429 Too Many Requests temporary ban, which can then cause a production downtime, which is... unacceptable?
Or did I misunderstand?

**Comment by matzmz** [comment]
I’m also looking to implement this solution to avoid the 429 Too Many Requests issue. It’s unclear to me whether the distribution service, even when specifying a tag that’s already cached, still requests the manifest from the source repository. If that’s how it works, it doesn’t seem to mitigate the 429 issue. Additionally, I’d like to understand how it handles the latest tag. Do you have any updates on this?

**Comment by milosgajdos** [comment]
> As I understand it, retrieving the manifest (latest version) will still count as a pull?

You need to clarify this with Docker Inc. 

This project implements a storage API in compliance with https://github.com/opencontainers/distribution-spec. That's it. What the commercial entities define as Pull/Push is not the goal of this project.

As for avoiding hitting Docker _Hub_ Registry API, yes, you can run the registry in proxy mode i.e. as a pull-through cache https://distribution.github.io/distribution/recipes/mirror/#run-a-registry-as-a-pull-through-cache
